core/src/main/java/com/linecorp/armeria/common/stream/FuseableStreamMessage.java

```diff
@@ -0,0 +1,280 @@
+            // (e.g., FuseableStreamMessage<T, U> is subtype of StreamMessage<U>.)
+            // So we don't know the first type when downcasting StreamMessage to FuseableStreamMessage.
+            final FuseableStreamMessage<?, T> cast = (FuseableStreamMessage<?, T>) source;
+            this.source = cast.source;
+
+            // Extract source functions and fuse them with the function
+            this.function = cast.function.and(function);
+        } else {
+            this.source = (StreamMessage<Object>) source;
+            this.function = (MapperFunction<Object, U>) function;
+        }
+    }
+
+    @VisibleForTesting
+    StreamMessage<Object> upstream() {
+        return source;
+    }
+
+    @Override
+    public boolean isOpen() {
+        return source.isOpen();
+    }
+
+    @Override
+    public boolean isEmpty() {
+        return source.isEmpty();
+    }
+
+    @Override
+    public long demand() {
+        return source.demand();
+    }
+
+    @Override
+    public CompletableFuture<Void> whenComplete() {
+        return source.whenComplete();
+    }
+
+    @Override
+    public void subscribe(Subscriber<? super U> subscriber, EventExecutor executor) {
+        subscribe(subscriber, executor, EMPTY_OPTIONS);
+    }
+
+    @Override
+    public void subscribe(Subscriber<? super U> subscriber, EventExecutor executor,
+                          SubscriptionOption... options) {
+        requireNonNull(subscriber, "subscriber");
+        requireNonNull(executor, "executor");
+        requireNonNull(options, "options");
+
+        source.subscribe(new FuseableSubscriber<U>(subscriber, function), executor, options);
+    }
+
+    @Override
+    public void abort() {
+        source.abort();
+    }
+
+    @Override
+    public void abort(Throwable cause) {
+        requireNonNull(cause, "cause");
+        source.abort(cause);
+    }
+
+    private static final class FuseableSubscriber<U> implements Subscriber<Object>, Subscription {
+
+        @SuppressWarnings("rawtypes")
+        private static final AtomicReferenceFieldUpdater<FuseableSubscriber, Subscription> upstreamUpdater =
+                AtomicReferenceFieldUpdater
+                        .newUpdater(FuseableSubscriber.class, Subscription.class, "upstream");
+
+        private final Subscriber<? super U> downstream;
+        private final MapperFunction<Object, U> function;
+
+        @Nullable
+        private volatile Subscription upstream;
+
+        FuseableSubscriber(Subscriber<? super U> downstream, MapperFunction<Object, U> function) {
+            requireNonNull(downstream, "downstream");
+            this.downstream = downstream;
+            this.function = function;
+        }
+
+        @Override
+        public void onSubscribe(Subscription subscription) {
+            requireNonNull(subscription, "subscription");
+            if (!upstreamUpdater.compareAndSet(this, null, subscription)) {
```

Comment:
I think we usually do not have this logic when we subscribe to a `StreamMessage`. Is there any reason that you put this? If so, we should apply the same approach to the other subscribers

Reply:
It is my mistake. 😅 It would be better to remove the validation from `onSubscribe()`.